File: src/main/java/org/micoli/php/symfony/messenger/navigation/MessengerGotoDeclarationHandler.java
package org.micoli.php.symfony.messenger.navigation;

import com.intellij.codeInsight.navigation.actions.GotoDeclarationHandler;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiElement;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.jetbrains.php.lang.psi.elements.*;
import java.util.Collection;
import org.jetbrains.annotations.Nullable;
import org.micoli.php.service.PhpUtil;
import org.micoli.php.symfony.messenger.service.MessengerService;

public class MessengerGotoDeclarationHandler implements GotoDeclarationHandler {

    @Override
    public PsiElement @Nullable [] getGotoDeclarationTargets(
            @Nullable PsiElement sourceElement, int offset, Editor editor) {

        if (sourceElement == null) {
            return null;
        }

        return handleDispatchNavigationToMessage(
                MessengerService.getInstance(sourceElement.getProject()), sourceElement);
    }

    public @Nullable PsiElement[] handleDispatchNavigationToMessage(
            MessengerService messengerService, PsiElement sourceElement) {

        if (!(sourceElement instanceof LeafPsiElement element)) {
            return null;
        }

        if (!messengerService.isDispatchMethod(element.getText())) {
            return null;
        }
        MethodReference methodRef = PsiTreeUtil.getParentOfType(sourceElement, MethodReference.class);

        if (methodRef == null) {
            return null;
        }

        String messageClassName = PhpUtil.getFirstParameterType(methodRef.getParameters());

        if (messageClassName == null) {
            return null;
        }

        Collection<Method> handlers =
                messengerService.findHandlersByMessageName(sourceElement.getProject(), messageClassName);

        if (handlers.isEmpty()) {
            return null;
        }

        return handlers.toArray(Method[]::new);
    }
}

File: src/main/kotlin/org/micoli/php/symfony/messenger/navigation/MessengerGotoDeclarationHandler.kt
package org.micoli.php.symfony.messenger.navigation

import com.intellij.codeInsight.navigation.actions.GotoDeclarationHandler
import com.intellij.openapi.editor.Editor
import com.intellij.psi.PsiElement
import com.intellij.psi.impl.source.tree.LeafPsiElement
import com.intellij.psi.util.PsiTreeUtil
import com.jetbrains.php.lang.psi.elements.MethodReference
import org.micoli.php.service.intellij.psi.PhpUtil.getFirstParameterType
import org.micoli.php.symfony.messenger.service.MessengerService

class MessengerGotoDeclarationHandler : GotoDeclarationHandler {
    override fun getGotoDeclarationTargets(
        sourceElement: PsiElement?,
        offset: Int,
        editor: Editor?
    ): Array<PsiElement?>? {
        if (sourceElement == null) {
            return null
        }

        val messengerService: MessengerService = MessengerService.getInstance(sourceElement.project)
        if (messengerService.configuration?.useNativeGoToDeclaration == true) {
            return null
        }
        return handleDispatchNavigationToMessage(messengerService, sourceElement)
    }

    fun handleDispatchNavigationToMessage(
        messengerService: MessengerService,
        sourceElement: PsiElement?
    ): Array<PsiElement?>? {
        if (sourceElement !is LeafPsiElement) {
            return null
        }

        if (!messengerService.isDispatchMethod(sourceElement.text)) {
            return null
        }
        val methodRef =
            PsiTreeUtil.getParentOfType(sourceElement, MethodReference::class.java) ?: return null

        val messageClassName = getFirstParameterType(methodRef.parameters) ?: return null

        val handlers = messengerService.findHandlersByMessageName(messageClassName)

        if (handlers.isEmpty()) {
            return null
        }

        return handlers.toTypedArray()
    }
}
